src/main/java/com/google/sps/ChatWebSocket.java

```diff
@@ -55,39 +82,109 @@ public void onOpen(Session session) throws IOException {
-    public void pongMessage(Session session, PongMessage msg) {
-        System.out.println("Pong message: " +
-            msg.getApplicationData().toString());
+
+    private void handleChatMessage(Session session, JSONObject messageData)
+            throws IOException {
+        JSONObject echoData = new JSONObject();
+        echoData.put(JSON_TYPE, MESSAGE_RECEIVE);
+        echoData.put(JSON_MESSAGE, messageData.getString(JSON_MESSAGE));
+        echoData.put(JSON_USER_ID, uid);
+
+        new Thread(){
+            public void run() {
+                DatastoreManager.getInstance().addMessage(chatRoomId, echoData);
+            }
+        }.start();
+
+        broadcastString(echoData.toString());
     }
 
-    @OnClose
-    public void onClose(Session session) throws IOException {
-        // WebSocket connection closes
+    private void handleMapDelete(Session session, JSONObject messageData)
+            throws IOException {
+        JSONObject echoData = new JSONObject(messageData,
+            new String[]{JSON_ID});
 
-        System.out.println("Socket " + session.getId() + " closed. " +
-            "Was authenticated: " + isAuthenticated);
+        //Delete marker
+        long markerId = messageData.getLong(JSON_ID);
+        DatastoreManager.getInstance().deleteMarker(chatRoomId, markerId);
 
-        if (!isAuthenticated) {
-            return;
+        //Broadcast deletion
+        echoData.put(JSON_TYPE, MAP_DELETE);
+
+        broadcastString(echoData.toString());
+    }
+
+    private void handleMapMessage(Session session, JSONObject messageData)
+            throws IOException {
+        JSONObject echoData = new JSONObject(
+            messageData, new String[]{JSON_TITLE, JSON_BODY, JSON_LATITUDE,
+                    JSON_LONGITUDE});
+
+        if (messageData.has(JSON_ID)) {
+            echoData.put(JSON_ID, messageData.getLong(JSON_ID));
+
+            new Thread(){
+                public void run() {
+                    DatastoreManager.getInstance()
+                                    .updateMarker(chatRoomId, echoData);
+                }
+            }.start();
+        }
+        else {
+            Key markerKey =
+                DatastoreManager.getInstance().newMarkerKey(chatRoomId);
+
+            new Thread(){
+                public void run() {
+                    DatastoreManager.getInstance()
+                                    .addMarker(markerKey, echoData);
+                }
+            }.start();
+
+            echoData.put(JSON_ID, markerKey.getId());
         }
 
-        WebSocketHandler.getInstance().removeSession(chatRoom, session);
+        echoData.put(JSON_TYPE, MAP_RECEIVE);
+
+        broadcastString(echoData.toString());
     }
 
-    @OnError
-    public void onError(Session session, Throwable throwable) {
-        // Do error handling here
+    @OnClose
+    public void onClose(Session session) throws IOException {
+        if (!isAuthenticated) {
+            return;
```

Comment:
maybe some error message?

Reply:
If the Session is not authenticated it does not necessarily require an error message. This happens if a websocket is connected and in the onOpen is not authenticated properly, it sends a close request and sends an error code there. This just verifies if the Session needs to be removed from the chatroom lists where messages are broadcasted.